### DISCUSSION TITLE
Example of exposing additional capabilities for runtime

### DISCUSSION THREAD

**Discussion by nazar-pc** [paritytech/substrate#9901] [Q&A] [answered]
There are certain capabilities exposed to offchain worker, but is there a reduced example that demonstrates exposing additional custom capabilities? Ideally not only for offchain worker, but runtime in general.

For background: I have a bunch of logic that is not possible to implement in runtime at the moment. If I could at least expose methods using runtime API with reduced API interface that can be upgraded when needed with runtime, that would be great, but that would still require access to various Substrate features and libraries that do not compile to WASM yes, so for that to work there must be a way to expose those additional APIs somehow.

**Comment by bkchr** [comment, marked as answer]
https://docs.rs/sc-executor/0.9.0/sc_executor/macro.native_executor_instance.html#with-custom-host-functions you mean something like this?

**Reply by nazar-pc** [reply]
It got refactored in https://github.com/paritytech/substrate/pull/9562, but it is probably what I'm looking for.

How would I use is though? Let's say I want to make a custom HTTP call or anything not available in the runtime by default really, some trivial example.

**Reply by bkchr** [reply]
For http we already provide functionalities. 

https://github.com/paritytech/substrate/blob/bf9683eee40f82cc4e01a05cd375b0e8bba3c8ef/frame/example-offchain-worker/src/lib.rs#L557

**Reply by nazar-pc** [reply]
I know that is it already provided, my goal is not HTTP client itself, I'm just looking for an example of implementing something custom that doesn't exist yet.

**Reply by bkchr** [reply]
When you just take what is there yet, then you can work based on this to create something that doesn't exist. Not sure there are tutorials for this. @NukeManDan maybe you know a tutorial?
